pennylane/compiler/qjit_api.py

```diff
@@ -459,8 +459,11 @@ def for_loop(lower_bound, upper_bound, step, loop_fn, *args):
 
     Args:
         lower_bound (int): starting value of the iteration index
-        upper_bound (int): (exclusive) upper bound of the iteration index
-        step (int): increment applied to the iteration index at the end of each iteration
+        upper_bound (int | None): (exclusive) upper bound of the iteration index. If ``None``,
+            provided ``lower_bound`` is used as the ``upper_bound`` and ``lower_bound`` is
+            considered to be ``0``.
+        step (int): increment applied to the iteration index at the end of each iteration.
+            Default is ``1``.
```

Comment:
@obliviateandsurrender I suggest we document this the same way NumPy documents [`arange()`](https://numpy.org/doc/stable/reference/generated/numpy.arange.html).

That is:

- The signature should show which args are optional
- there is a note outside the Args section showing the three ways it can be called
- The `Args` simply describe the behaviour of the three args (without trying to capture the three different ways it can be called).

It might also be worth changing the args to `start, stop, step` to match `arange`?

Reply:
I made all the changes.